feat: implementing async states

dapeventigan opened the pull request:
- added async dart that contains async enums/states
- added Redux's Wait in app_state
- implemented async in vm
- implemented async's enum/states in overview page which are current state, loading, and error

[app overview]
https://user-images.githubusercontent.com/112401492/227859978-e33738b5-3006-461d-9e77-81a62792b5fc.mp4

[updated app overview]
https://user-images.githubusercontent.com/112401492/227901604-0b6f2a8b-6054-41f5-9e58-f1d8cfb6bcf2.mp4

[updated app overview 2]
https://user-images.githubusercontent.com/112401492/227914881-b02afb69-d327-4d28-8699-eb40b77d0c45.mp4


[flutter analyze and dart]
![image](https://user-images.githubusercontent.com/112401492/227860225-c376bcde-46e5-4c68-bd87-f28c6aefa7a2.png)

dapeventigan committed:
feat: implementing async states

- added async dart that contains async enums/states
- added Redux`s Wait in app_state
= implemented async in vm
- implemented async`s enum/states in overview page which are current state, loading, and error

dapeventigan committed:
feat: implementing async states

- turned pokemon action key into constant

dapeventigan committed:
feat: feat: implementing async states

- wrapped snackbar into a separate function
- fixed error message

dapeventigan committed:
feat: implementing async states

- fixed snackbar
- fixed error message

dapeventigan committed:
feat: implementing async states

- fixed snackbar by implementing addPostFrameCallback outside the void function

shawn-ffufm commented on pokedex_async_redux/lib/features/pokemon_overview_page.dart:
```diff
@@ -10,25 +11,50 @@ class PokemonOverviewPage extends StatelessWidget {
-  final List<Pokemon> pokemons;
+  final Async<List<Pokemon>> pokemons;
+
+  void _showErrorMessageSnackbar(BuildContext context, String message) {
```
please have this function placed below build method 🙂

shawn-ffufm commented on pokedex_async_redux/lib/features/pokemon_overview_page.dart:
```diff
@@ -10,25 +11,50 @@ class PokemonOverviewPage extends StatelessWidget {
+  final Async<List<Pokemon>> pokemons;
+
+  void _showErrorMessageSnackbar(BuildContext context, String message) {
+    WidgetsBinding.instance.addPostFrameCallback(
```
see cleo's pr for reference on how she segregated the snackbar implementation with the postFrameCallback 🙂

shawn-ffufm commented on pokedex_async_redux/lib/features/pokemon_overview_page.dart:
```diff
@@ -10,25 +11,50 @@ class PokemonOverviewPage extends StatelessWidget {
+    WidgetsBinding.instance.addPostFrameCallback(
+      (_) => ScaffoldMessenger.of(context).showSnackBar(
+        SnackBar(
+          content: Text(message),
```
no need for extra trailing comma here 🙂

dapeventigan committed:
feat: implementing async states

- placed snackbar below build
- fixed trailing commas

shawn-ffufm commented on pokedex_async_redux/lib/features/pokemon_overview_page.dart:
```diff
@@ -10,25 +11,50 @@ class PokemonOverviewPage extends StatelessWidget {
+
+  void _showErrorMessageSnackbar(BuildContext context, String message) {
+    ScaffoldMessenger.of(context).showSnackBar(
+      SnackBar(
```
no need for extra trailing comma 🙂

shawn-ffufm commented on pokedex_async_redux/lib/features/pokemon_overview_page.dart:
```diff
@@ -10,25 +11,50 @@ class PokemonOverviewPage extends StatelessWidget {
+  void _showErrorMessageSnackbar(BuildContext context, String message) {
+    ScaffoldMessenger.of(context).showSnackBar(
+      SnackBar(
+        content: Text(message),
```
same here

shawn-ffufm commented on pokedex_async_redux/lib/features/pokemon_overview_page.dart:
```diff
@@ -10,25 +11,50 @@ class PokemonOverviewPage extends StatelessWidget {
-            itemBuilder: (context, index) {
-              final pokemon = pokemons[index];
-              return PokemonCard(pokemon: pokemon);
+      home: Builder(
```
why wrap this in a Builder? seems no need for this, u can maximize using the build method's provided context 🙂

shawn-ffufm commented on pokedex_async_redux/lib/features/pokemon_overview_page.dart:
```diff
@@ -10,25 +11,50 @@ class PokemonOverviewPage extends StatelessWidget {
+          body: pokemons.when(
+            loading: () => const Center(
+              child: CircularProgressIndicator(
+                color: appDefaultColor,
```
no need for extra trailing comma here

shawn-ffufm commented on pokedex_async_redux/lib/features/pokemon_overview_page.dart:
```diff
@@ -10,25 +11,50 @@ class PokemonOverviewPage extends StatelessWidget {
+            ),
+            error: (errorMessage) {
+              WidgetsBinding.instance.addPostFrameCallback(
+                (_) => _showErrorMessageSnackbar(context, errorMessage!),
```
no need for extra trailing comma here

shawn-ffufm commented on pokedex_async_redux/lib/features/pokemon_overview_page.dart:
```diff
@@ -10,25 +11,50 @@ class PokemonOverviewPage extends StatelessWidget {
+                (_) => _showErrorMessageSnackbar(context, errorMessage!),
+              );
+              return const Center(
+                child: Text(noPokemonAvailable),
```
no need for extra trailing comma here

shawn-ffufm commented on pokedex_async_redux/lib/state/action/action.dart:
```diff
@@ -0,0 +1,14 @@
+import 'package:async_redux/async_redux.dart';
+import 'package:pokedex_async_redux/state/app_state.dart';
+
+abstract class LoadingAction extends ReduxAction<AppState> {
```
you forgot to add a concise documentation 😝

remember, this is mandtoy right? 😉

shawn-ffufm commented on pokedex_async_redux/lib/state/action/pokemon_actions.dart:
```diff
@@ -2,11 +2,16 @@ import 'dart:async';
 /// Setting of pokemons from pokemon api
-class GetPokemonsAction extends ReduxAction<AppState> {
+class GetPokemonsAction extends LoadingAction {
+  static const key = pokemonActionKey;
```
this is already a constant right? therefore no need to create a separate one for this 😉

shawn-ffufm commented on pokedex_async_redux/lib/utilities/constants.dart:
```diff
@@ -8,3 +11,8 @@ const idIndex = 6;
 const indexValue = '[index]';
 const pokemonImageUri =
     "https://raw.githubusercontent.com/PokeAPI/sprites/master/sprites/pokemon/other/home/[index].png";
+const noInternetText = "No Internet Connection";
```
do u need this?

shawn-ffufm commented on pokedex_async_redux/lib/utilities/constants.dart:
```diff
@@ -8,3 +11,8 @@ const idIndex = 6;
 const pokemonImageUri =
     "https://raw.githubusercontent.com/PokeAPI/sprites/master/sprites/pokemon/other/home/[index].png";
+const noInternetText = "No Internet Connection";
+const errorCircleIcon = Icons.error_outline;
```
also this?

shawn-ffufm commented on pokedex_async_redux/lib/utilities/constants.dart:
```diff
@@ -8,3 +11,8 @@ const idIndex = 6;
     "https://raw.githubusercontent.com/PokeAPI/sprites/master/sprites/pokemon/other/home/[index].png";
+const noInternetText = "No Internet Connection";
+const errorCircleIcon = Icons.error_outline;
+const noPokemonAvailable = 'No Pokemon Available';
```
Pokemons - should be plural

shawn-ffufm commented on pokedex_async_redux/lib/utilities/constants.dart:
```diff
@@ -8,3 +11,8 @@ const idIndex = 6;
+const errorCircleIcon = Icons.error_outline;
+const noPokemonAvailable = 'No Pokemon Available';
+const errorMessage = 'Something went wrong!';
+const pokemonActionKey = 'get pokemon action';
```
see feedback about this

shawn-ffufm reviewed (changes requested):
see pr feedback 🙂

dapeventigan committed:
feat: implementing async states

- removed Builder and changed the overview page MaterialApp and Scaffold implementation
- fixed trailing commas
- added comment documentation in LoadingAction

dapeventigan committed:
feat: implementing async states

- minor fix on constant values

shawn-ffufm commented on pokedex_async_redux/lib/features/pokemon_overview_page.dart:
```diff
@@ -10,28 +11,40 @@ class PokemonOverviewPage extends StatelessWidget {
+      ),
+      body: pokemons.when(
+        loading: () => const Center(
+          child: CircularProgressIndicator(color: appDefaultColor),
```
no need for extra trailing comma 🙂

shawn-ffufm commented on pokedex_async_redux/lib/features/pokemon_overview_page.dart:
```diff
@@ -10,28 +11,40 @@ class PokemonOverviewPage extends StatelessWidget {
         ),
-        body: Center(
+        error: (errorMessage) {
+          WidgetsBinding.instance.addPostFrameCallback((_) => _showErrorMessageSnackbar(context, errorMessage!));
```
use null coalescing and use `emptyString` constant as its fallback value instead of using non-null enforce here to be safer 🙂

shawn-ffufm reviewed (changes requested):
see minor pr feedback 🙂

dapeventigan committed:
feat: implementing async states

- fixed trailing commas
- used null coalescing on error message instead of non-null enforce

shawn-ffufm reviewed (approved):
lgtm
